### ISSUE TITLE
prisma seems to preserve more detail

### ISSUE THREAD

**Issue by xlvector** [jcjohnson/fast-neural-style#58]
I find prisma seems to preserve more details when do style transfer. Following is an example:

origin image

![free](https://cloud.githubusercontent.com/assets/100203/19834713/5f8a0660-9ea6-11e6-8769-aa2677a44e3d.jpg)

prisma use udnie style

![free_udnie](https://cloud.githubusercontent.com/assets/100203/19834719/7b4ec336-9ea6-11e6-8086-142bcfe4f2d1.jpg)

fast-neural-style results use udnie style

![th_free_udnie](https://cloud.githubusercontent.com/assets/100203/19834725/91d5e40e-9ea6-11e6-944a-34e8df35343e.jpg)

There are following differences:
1. udnie style have many red, yellow, orange color, which also appears in fast-neural-style result, but does not exist in prisma result
2. prisma result have smooth sky
3. prisma preserve more details in bridge and tree

I do not know how does prisma achieve this, I have already tune many hyper parameters during training.

**Comment by htoyryla** [comment]
I guess Prisma is copying color from the original image here, which could explain some of the differences.

Somehow I feel that while both results are interesting, neither really captures much of the original style. Fast-neural-style fills the picture with a colored mesh which indeed captures the forms in the content image, using colors from the style image but not really resembling the shapes, their scale and feeling of the original style. 

I may be mistaken but I think the iterative neural-style was better in real style transfer. Fast-neural-style is a great tool for creating styles but these style tend to look very different from the originals. I have had similar experiences with texture_nets, with which I experiments for days trying get to the style reproduced in more or less the original scale, until I gave up and moved to something else. I have not yet tried the same with fast-neural-style.

By the way, to me it looks like the new style transfer methods can easily fill the canvas with decorating stylistic details, but the opposite, simplifying, is difficult. And yet, much of art is about simplifying what you see and capturing it in an image. Prisma, I think, is closer, but not exactly what I am after.

**Comment by htoyryla** [comment]
Here's the result from my newly trained model after 40k iterations. The "colored mesh" did not spread out throughout the sky, as I feared, but in fact retreated and the overall look improved. 

In this image I'd point out the white "ghosts" behind the bridge and the buildings on the left. Here they blend quite well into the background but in my experiments I've seen images almost totally dominated by such "ghost" shapes, especially in the sky, and especially with higher style weights. 

![undie_ny-test000h](https://cloud.githubusercontent.com/assets/15064373/19836549/8876e4d4-9eac-11e6-942c-de3ecac9ba29.png)

**Comment by htoyryla** [comment]
I repeated the training with content_weight=1, style_weight=3, up to 40k iterations. Here's the resulting image. I then wrote a script to copy original colors to an image (https://gist.github.com/htoyryla/147f641f2203ad01b040f4b568e98260) and using it made the second image.

I think it would be possible to get even closer to the Prisma look by still finetuning the weights, and to get a touch of detail, blend the different channels of the original image in a suitable mix into resulting image, instead of simply copying the color information.

Output from fast_neural_style.lua:
![undie_ny-test001z](https://cloud.githubusercontent.com/assets/15064373/19837808/7f5183d2-9eca-11e6-8438-94acf056174b.png)

Output from original_colors.lua:
![t001z](https://cloud.githubusercontent.com/assets/15064373/19837810/8ba58854-9eca-11e6-9f73-9a4467a86887.png)

**Comment by jcjohnson** [comment]
Wow, nice work @htoyryla! I think your Udnie model is better than mine :)

In general I don't think that Prisma is doing anything fundamentally different from fast-neural-style; I think they have just spent a lot of time and effort carefully tuning the hyperparameters of their models to give nice effects. I think they also do some post-processing to blend the raw neural-net output with the content image; there are a lot of different ways to blend images, and I think they also tune the post-processing per style to make sure their results are nice.

**Comment by htoyryla** [comment]
_"I think they also do some post-processing to blend the raw neural-net output with the content image; there are a lot of different ways to blend images, and I think they also tune the post-processing per style to make sure their results are nice."_

That's exactly what I was thinking when I wrote about blending. I simply copied the Y channel, but if one wants a touch of detail then I think one could blend some of the other channels. And the optimum way to do this is likely to be specific to style.

I used my own dataset consisting of 2500 photos, places and landscapes. I have noticed that using it can give quite different results from mscoco. I'll check now the same training but using mscoco.

**Comment by jcjohnson** [comment]
> I used my own dataset consisting of 2500 photos, places and landscapes. I have noticed that using it can give quite different results from mscoco. I'll check now the same training but using mscoco.

Interesting; I have only tried training with COCO but I'm pretty sure the training images are important. I think the number of training images is also important; In Dmitry's Instance Normalization paper (https://arxiv.org/abs/1607.08022) he mentions that his best results were trained with only 16 content images. I haven't done much experimentation with training sets, but this seems to be an important area for exploration.

**Comment by htoyryla** [comment]
Have now trained using COCO but otherwise the same parameters. Different but not too different. Looks almost as if I had used a bit higher style weight. 

From fast-neural-style.lua:

![undie_ny-mscoco001z](https://cloud.githubusercontent.com/assets/15064373/19839403/989b6122-9ee9-11e6-928c-49250e4a487d.png)

After original_color.lua:

![t002z](https://cloud.githubusercontent.com/assets/15064373/19839408/ae7488c0-9ee9-11e6-9ece-4358bb71f711.png)

**Comment by xlvector** [comment]
The results looks awesome! Thanks!

I will try to do more works on post-processing.

**Comment by xlvector** [comment]
@jcjohnson I have tried to train with 200K images from coco, MIT space, and imageNet. Seems does not get better results. I will try this again later.

**Comment by xlvector** [comment]
@jcjohnson what is your parameter for the_wave style. I seems can not reproduce your results with parameter in print_options.lua.

Following is my result:

![free_coco_wave7](https://cloud.githubusercontent.com/assets/100203/19881813/d038fdaa-a043-11e6-9a35-7e899fa24d57.jpg)

Your results

![th_free_wave](https://cloud.githubusercontent.com/assets/100203/19881820/deeb1c0c-a043-11e6-8df1-ac0dd2f488c7.jpg)

prisma

![free_wave](https://cloud.githubusercontent.com/assets/100203/19881823/e99fc2ce-a043-11e6-99a1-123c1d260f74.jpg)

**Comment by jcjohnson** [comment]
My wave model does not use instance norm, so you should set `-use_instance_norm 0` if you want to duplicate my results.

**Comment by piteight** [comment]
Is there a way to change parameters manually inside a model? Something like selecting weights that interests us, and changing their values? This script uses vgg16 model, but I also have not trained vgg19 model which gave me better results in slow neural style from [fzliu](https://github.com/fzliu/style-transfer). Can I switch to training vgg19  model with your script?
Here's my results in vgg16:)
trained image 
![stasio](https://cloud.githubusercontent.com/assets/22580151/19925756/a5b0cc80-a0f1-11e6-8555-4b83f16cc52a.jpg)


`th train.lua -h5_file ../Baz.h5 -style_image ../stasio.jpg -style_image_size 256 -content_weights 1.0 -style_weights 5.0 -checkpoint_name stasio -gpu 0 -use_cudnn 1 -backend cuda -batch_size 2 -checkpoint_every 100`

![ay](https://cloud.githubusercontent.com/assets/22580151/19925728/7ab48b20-a0f1-11e6-9d5b-ef039b2b7316.jpg)

`th train.lua -h5_file ../Baz.h5 -style_image ../stasio.jpg -style_image_size 300 -content_weights 3.0 -style_weights 1.0 -checkpoint_name ztasio -gpu 0 -use_cudnn 1 -backend cuda -batch_size 2 -checkpoint_every 100 -arch c9s1-16,d32,d64,R64,R64,R64,R64,R64,U2,U2,c9s1-3 
`
![az](https://cloud.githubusercontent.com/assets/22580151/19925793/c2226824-a0f1-11e6-9016-0a29d13bd4a0.jpg)
`th train.lua -h5_file ../Baz.h5 -style_image ../stasio.jpg -style_image_size 300 -content_weights 3.0 -style_weights 5.0 -checkpoint_name Xtasio -gpu 0 -use_cudnn 1 -backend cuda -batch_size 2 -checkpoint_every 100 -arch c9s1-16,d32,d64,R64,R64,R64,R64,R64,U2,U2,c9s1-3 -max_train 1
`
![azz](https://cloud.githubusercontent.com/assets/22580151/19925851/0e3312ae-a0f2-11e6-8642-a8fcbba07192.jpg)

`th train.lua -h5_file ../Baz.h5 -style_image ../stasio.jpg -style_image_size 300 -content_weights 0.5 -style_weights 8.0 -checkpoint_name Ytasio -gpu 0 -use_cudnn 1 -backend cuda -batch_size 2 -checkpoint_every 100 -arch c9s1-16,d32,d64,R64,R64,R64,R64,R64,U2,U2,c9s1-3`
![azzz](https://cloud.githubusercontent.com/assets/22580151/19925881/35c69c8c-a0f2-11e6-8c2e-3e7c43666489.jpg)

The first model is the best, but It have some glitches that I want to erease in future training sessions.
Everything was trained with COCO 40k images.

**Comment by htoyryla** [comment]
I see you have copied the arch c9s1-16,d32,d64,R64,R64,R64,R64,R64,U2,U2,c9s1-3 from my comment. As commented by @jcjohnson in another thread, that might be a poor choice, and perhaps one should add a conv layer between the two U2 layers (even if for me it worked without).

**Comment by piteight** [comment]
Yes, I thought I would give it a try, to see different methods than only changing weights of style and content. The second version is quite good, because of the sky,  except this noise pattern. It worket good for chicago.jpg, 
![chicagoz](https://cloud.githubusercontent.com/assets/22580151/19928909/0bafe796-a101-11e6-872b-2c79f7569ecc.jpg)
but in example with person, the result was very poor:
![ztkaw](https://cloud.githubusercontent.com/assets/22580151/19928975/55ca2648-a101-11e6-828c-e62d96569a5c.jpg)


first example gave me this output:
![out](https://cloud.githubusercontent.com/assets/22580151/19928945/353645ec-a101-11e6-9c6e-592eab3fd481.png)
![tkaw](https://cloud.githubusercontent.com/assets/22580151/19928980/5a509e7c-a101-11e6-9232-dada9aaa6441.jpg)

I will try putting the conv layer as You mentioned :)

**Comment by universewill** [comment]
@htoyryla can u make your dataset and pretrained models' parameters available?

**Comment by htoyryla** [comment]
> @htoyryla can u make your dataset and pretrained models' parameters available?

After almost three years of doing other things, no, sorry.
